docs: clearly deprecate Harness in the testing how-tos (#1508)
canonical/operator: docs/howto/get-started-with-charm-testing.md

 
You will notice that the starting point is typically always an event. A charm doesn't do anything unless it's being run, and it is only run when an event occurs. So there is *always* an event context to be mocked. This has important consequences for the unit-testing framework, as we will see below.

> ### The framework

Reviewer: Suggestion for consistency with the following heading ("Understanding the test framework")

```suggestion
### The test framework
```

In the charming world, unit testing means state-transition testing.

> See more [`ops.testing`](https://ops.readthedocs.io/en/latest/reference/ops-testing.html)

> `State` is the 'mocker' for most inputs and outputs you will need. Where a live charm would gather its input through context variables and calls to the Juju api (by running the hook tools), a charm under unit test will gather data via a mocked backend managed by the testing framework. Where a live charm would produce output by writing files to a filesystem, `Context` and `Container` expose a mock filesystem the charm will be able to interact with without knowing the difference. More specific outputs, however, will need to be mocked individually.

Reviewer 2: Style tweaks while we're here:

```suggestion
`State` is the 'mocker' for most inputs and outputs you will need. Where a live charm would gather its input through context variables and calls to the Juju API (by running the hook tools), a charm under unit test will gather data using a mocked backend managed by the testing framework. Where a live charm would produce output by writing files to a filesystem, `Context` and `Container` expose a mock filesystem the charm will be able to interact with without knowing the difference. More specific outputs, however, will need to be mocked individually.
```

A typical test will look like this:
 
- set things up:
  - set up the charm and its metadata
  - set up the context

>   - mock any 'output' callable that you know would misfire or break (e.g. a system call -- you don't want a unit test to reboot your laptop)

Reviewer 2: ```suggestion
  - mock any 'output' callable that you know would misfire or break (for example, a system call -- you don't want a unit test to reboot your laptop)
```

  - set up the Juju state in which the event will fire, including config and relation data

>  - **mock an event**

Reviewer 2: I wonder if we can improve this "mock an event" language to use "Scenario"-specific terms, like "trigger an event using `Context.run`". We're not really "mocking" an event.

 - get the output
 - run assertions on the output

> Obviously, other flows are possible; for example, where you unit test individual charm methods without going through the whole event context setup, but this is the characteristic one.

> ### Understanding the test framework

Reviewer: ```suggestion
### Understanding the testing framework
```

def test_pebble_ready_writes_config_file():
    """Test that on pebble-ready, a config file is written"""
    ctx = testing.Context(MyCharm)
    # If you want to mock charm config:
    config = {'foo': 'bar'}
    # If you want to mock charm leadership:

>     leader = True

Reviewer 2: Might be more natural to move `config={'foo': 'bar'}` and `leader=True` to inline in the `State()` call.
